File: slidingWindow/longestSubstring.ts



function longestSubstring(str: string) {

    if (str.length <= 1) {
        return str.length;
    }

    let leftPointer = 0;
    let longestSubstringCount = 0;

    const seenCharacters: { [key: string]: number } = {};

    for (let rightPointer = 0; rightPointer < str.length; rightPointer++) {
        const currentCharacter = str[rightPointer];
        const seenCharacterIndex = seenCharacters[currentCharacter];

        // character found in hash map

        if (seenCharacterIndex >= leftPointer) {
            leftPointer = seenCharacterIndex + 1;
        }

        seenCharacters[currentCharacter] = rightPointer;

        longestSubstringCount = Math.max(longestSubstringCount, rightPointer - leftPointer + 1)


    }

    return longestSubstringCount;
}

longestSubstring("abcbdabc");

File: slidingWindow/longestSubstring.go
package main

import (
	"fmt"
	"math"
)

func longestSubstring(str string) int {
	if len(str) <= 1 {
		return len(str)
	}

	leftPointer := 0
	longestSubstringCount := 0

	seenCharactersMap := map[string]int{}

	for rightPointer := 0; rightPointer < len(str); rightPointer++ {

		currentCharacter := string(str[rightPointer])
		seenCharacterIndex := seenCharactersMap[currentCharacter]

		if seenCharacterIndex >= leftPointer {
			leftPointer = seenCharacterIndex + 1
		}

		seenCharactersMap[currentCharacter] = rightPointer
		longestSubstringCount = int(math.Max(float64(longestSubstringCount), float64(rightPointer-leftPointer+1)))
	}
	return longestSubstringCount

}

func main() {
	longestSubstringCount := longestSubstring("abcdabc")

	fmt.Printf("The longest substring count is %d", longestSubstringCount)
}
